### PR TITLE
Cairo bindings

### PR DESCRIPTION
**Pull request by karol-bisztyga**
Python bindings for https://github.com/software-mansion-labs/cairo/

### CODE REVIEW

File: install_cairo.sh
Comment (karol-bisztyga) [inline install_cairo.sh:15]: this can be removed once https://github.com/software-mansion-labs/cairo.git works properly

File: tests/data/contracts.py
Comment (MaksymilianDemitraszek) [inline tests/data/contracts.py:141]: I would put cairo1 contracts in a separate file to avoid confusion

File: tests/data/contracts.py
Comment (MaksymilianDemitraszek) [inline tests/data/contracts.py:233]: You can use cairo1_test_suite_parser to run those tests (you can try using roll cheatcode inside). An example how to do it is `tests/integration/cairo1/parser_test.py`

Comment (karol-bisztyga) [reply tests/data/contracts.py:233]: I think we should not do this here. It's good to have tests that only check if the compilation is OK. I can do this in a separate test that uses both the parser and the compiler.

File: tests/data/contracts.py
Comment (karol-bisztyga) [reply tests/data/contracts.py]: [move cairo 1 contracts to a separate file](https://github.com/software-mansion/protostar/pull/1316/commits/1c44c3a48da4ed718c609b7cdf1d06019fa3288f)

File: tests/data/contracts.py
Comment (karol-bisztyga) [reply tests/data/contracts.py]: [add tests for compiler+parser](https://github.com/software-mansion/protostar/pull/1316/commits/aa9a39455f7c1641671744f4717804bf86379de7)

Comment (MaksymilianDemitraszek) [reply tests/data/contracts.py]: I am okay with that, It was just a hint 😉

**Comment by karol-bisztyga** [comment]
Please, confirm whether the CI fix is valid 🙏

File: tests/integration/cairo_compiler/tests_compilation_test.py
Comment (Arcticae) [inline tests/integration/cairo_compiler/tests_compilation_test.py:43]: This thing is redundant. If you pass a `Path` object to the function, it'll automatically read the file, and if you pass a `str`, it will get the code from the string.

File: tests/integration/cairo_compiler/contracts_compilation_test.py
Comment (Arcticae) [inline tests/integration/cairo_compiler/contracts_compilation_test.py:43]: no need for this logic

File: install_cairo.sh
Comment (Arcticae) [reply install_cairo.sh:19]: I mean adding it normally through `poetry add maturin`

**Review by Arcticae — CHANGES_REQUESTED** [review]
It's 🔥 that it works, really hyped about this one! Good job!

Some style changes and nitpicking left, that i think should be discussed (or explained) either in the code or here in review.

File: tests/integration/cairo1/parser_test.py
Comment (kasperski95) [inline tests/integration/cairo1/parser_test.py:14]: Consider renaming  "compiled_test_suite.json" and moving it next to the source code.

File: tests/data/cairo1_contracts.py
Comment (karol-bisztyga) [reply tests/data/cairo1_contracts.py:1]: I don't mind where they are put and read honestly.

Yes, they are copied and it's done on purpose because it's better (easier) to have exactly the same contracts here and in the cairo repository so if anything goes wrong we can debug it. At least for now, they're constantly changing different features and it's really important to be sure that we do as many things as possible exactly the same as on their side.

> Testing the compiler implementation is not something we'd like to do.

I think it doesn't matter that much what contracts here we're going to compile, what matters is that exactly the same contracts work with the original compiler. That's how we don't test the compiler itself but the bindings.

**Comment by karol-bisztyga** [comment]
I don't like this entire installation process but I couldn't come up with something better (considering CI). I'm going to try to improve this but nevertheless, I think we should just leave it as is (roughly ok) and create a task for fixing it.

File: install_cairo.sh
Comment (karol-bisztyga) [reply install_cairo.sh:13]: Yes, afaik the work for this is scheduled or in progress atm

File: tests/data/cairo1_contracts.py
Comment (karol-bisztyga) [reply tests/data/cairo1_contracts.py:1]: As for the place for the contracts, this is done similarily to places like these:
```
tests/integration/data_transformer/contracts.py
tests/data/contracts.py

```

File: tests/integration/cairo_compiler/tests_compilation_test.py
Comment (karol-bisztyga) [reply tests/integration/cairo_compiler/tests_compilation_test.py:43]: Sorry, but I don't understand :/ What function do you mean?

Comment (kasperski95) [reply tests/integration/cairo_compiler/tests_compilation_test.py:43]: `ProtostarFixture::create_files`/`ProtostarFixture::create_contracts`

File: install_cairo.sh
Comment (karol-bisztyga) [reply install_cairo.sh:11]: Hmm, most of the commands are pretty easy I think, going around directories, cloning the repo, etc. I bet the biggest problem is maturin. I understand it works like this, you log into the poetry shell, then go to the other repo directory, perform `maturin develop` and it installs the binding, but what exactly it does underneath, for now, I'm not sure, sorry.

File: tests/integration/cairo_compiler/contracts_compilation_test.py
Comment (karol-bisztyga) [reply tests/integration/cairo_compiler/contracts_compilation_test.py:63]: This is a simple test of the compilation process, but I agree the test itself could be more precise. We should probably check the results with the hardcoded compiled programs.

File: tests/integration/cairo_compiler/compilator_and_parser_test.py
Comment (karol-bisztyga) [reply tests/integration/cairo_compiler/compilator_and_parser_test.py:21]: I understand you mean I could use `build`? In such a case, I'd await the function anyway right away so there's no difference to me unless you meant something different.

File: tests/integration/cairo_compiler/compilator_and_parser_test.py
Comment (karol-bisztyga) [reply tests/integration/cairo_compiler/compilator_and_parser_test.py]: [move bindings to a sep file](https://github.com/software-mansion/protostar/pull/1316/commits/13308141b1261f1a66f63e6e9d771cdc05d563ae) I moved the functions to a separate place so they're nicely typed, I know I still use `# pyright ignore` but they're in one place and if we come up with a better solution than just suppressing the checker, it will be easier to implement.

File: tests/integration/cairo_compiler/tests_compilation_test.py
Comment (karol-bisztyga) [reply tests/integration/cairo_compiler/tests_compilation_test.py]: [extract fixture to a sep file](https://github.com/software-mansion/protostar/pull/1316/commits/25d34a76296e38402292118940d46c8917b2c850)

File: tests/integration/cairo_compiler/compilator_and_parser_test.py
Comment (karol-bisztyga) [reply tests/integration/cairo_compiler/compilator_and_parser_test.py]: [extract fixture to a sep file](https://github.com/software-mansion/protostar/pull/1316/commits/25d34a76296e38402292118940d46c8917b2c850)

File: tests/integration/cairo_compiler/contracts_compilation_test.py
Comment (karol-bisztyga) [reply tests/integration/cairo_compiler/contracts_compilation_test.py]: [extract fixture to a sep file](https://github.com/software-mansion/protostar/pull/1316/commits/25d34a76296e38402292118940d46c8917b2c850)

File: install_cairo.sh
Comment (karol-bisztyga) [reply install_cairo.sh]: [add maturing through poetry](https://github.com/software-mansion/protostar/pull/1316/commits/2d89bdef4134f11fca539e8b836b2bf20a222d0f)

**Comment by karol-bisztyga** [comment]
I significantly refactored tests in [tests refactor](https://github.com/software-mansion/protostar/pull/1316/commits/77c11e61e8d2afd557776ef221436bb93ed4bbfa), I think they're easier to read now (I admit they were a bit messy). I'm resolving all the conversations regarding this then.

File: tests/data/cairo1_contracts.py
Comment (karol-bisztyga) [reply tests/data/cairo1_contracts.py]: [move contracts' contents to separate files](https://github.com/software-mansion/protostar/pull/1316/commits/7a524ce16343cda8c9747fc3bbe5449a31023aa0)

File: install_cairo.sh
Comment (karol-bisztyga) [reply install_cairo.sh:8]: right, the `CFLAGS` stuff is mac-specific

File: install_cairo.sh
Comment (karol-bisztyga) [reply install_cairo.sh]: [try new master](https://github.com/software-mansion/protostar/pull/1316/commits/0ae741ca086d91e872624990b1021d849718558e) working now with the current master

File: install_cairo.sh
Comment (karol-bisztyga) [reply install_cairo.sh]: [install bindings as a poe script](https://github.com/software-mansion/protostar/pull/1316/commits/e02df263886f09ce5c1125667cefc00deb855f40)

Comment (karol-bisztyga) [reply install_cairo.sh]: I mostly agree.

One thing I don't think is worth doing is checking for the tools like `rustup`. We're using the command just like that so if rust is not installed, you'll probably get something like `rustup - command not found` which is already pretty descriptive.

File: tests/integration/cairo_compiler/compilator_and_parser_test.py
Comment (karol-bisztyga) [reply tests/integration/cairo_compiler/compilator_and_parser_test.py]: https://github.com/software-mansion/protostar/issues/1434

File: install_cairo.sh
Comment (karol-bisztyga) [reply install_cairo.sh]: I mean m1-specific

File: install_cairo.sh
Comment (karol-bisztyga) [reply install_cairo.sh]: [add a note to the docs about installing bindings](https://github.com/software-mansion/protostar/pull/1316/commits/6545be744f9c4681d61a133c1ac01756bb3e758f)

File: tests/integration/cairo_compiler/compilator_and_parser_test.py
Comment (karol-bisztyga) [reply tests/integration/cairo_compiler/compilator_and_parser_test.py]: [add a todo with an issue no](https://github.com/software-mansion/protostar/pull/1316/commits/143477c41a64b2528a5f3be0602ef15f25ec4670)

File: install_cairo.sh
Comment (karol-bisztyga) [reply install_cairo.sh]: [poetry install depends on whether it works on m1](https://github.com/software-mansion/protostar/pull/1316/commits/a2c08590a21d18098ac4995be31a2922ae27c27a)

File: scripts/install_cairo_bindings.sh
Comment (karol-bisztyga) [inline scripts/install_cairo_bindings.sh:13]: this whole section should probably be deleted

File: .github/workflows/checks.yml
Comment (karol-bisztyga) [reply .github/workflows/checks.yml]: Yes, but I'm not sure if it's possible now (maybe it is). `maturin build` produces a wheel and maybe we could publish that wheel and install it via poetry and then it probably would be cached like other packages.

Long story short, I'd stay with what we have now and address this later, I think the whole installation process will change.

https://github.com/software-mansion/protostar/issues/1436

Comment (karol-bisztyga) [reply .github/workflows/checks.yml]: Please resolve, if that's ok.

File: protostar/cairo/cairo_bindings.py
Comment (Arcticae) [inline protostar/cairo/cairo_bindings.py:1]: wow nice, that's exactly what i meant, good job!

File: tests/integration/cairo_compiler/prepare_files_fixture.py
Comment (Arcticae) [inline tests/integration/cairo_compiler/prepare_files_fixture.py:42]:
```suggestion
            {str(path): contents for (path, contents) in files.values()}
```

File: tests/integration/cairo_compiler/prepare_files_fixture.py
Comment (Arcticae) [inline tests/integration/cairo_compiler/prepare_files_fixture.py:16]: You can use enum with string values here

Comment (Arcticae) [reply tests/integration/cairo_compiler/prepare_files_fixture.py:16]: And all the `if` statements from `prepare_files` can be removed then :)

File: tests/integration/cairo_compiler/prepare_files_fixture.py
Comment (Arcticae) [inline tests/integration/cairo_compiler/prepare_files_fixture.py:49]:
```suggestion
        for (path, _) in files.values():
            assert path.exists()
```

Comment (Arcticae) [inline tests/integration/cairo_compiler/prepare_files_fixture.py:49]:
```suggestion
        for (path, _) in files.values():
            assert path.exists()
```

File: tests/integration/cairo_compiler/prepare_files_fixture.py
Comment (Arcticae) [inline tests/integration/cairo_compiler/prepare_files_fixture.py:10]:
```suggestion
class RequestedFile(Enum):
```

Comment (Arcticae) [inline tests/integration/cairo_compiler/prepare_files_fixture.py:10]:
```suggestion
class RequestedFile(Enum):
```

File: tests/integration/cairo_compiler/prepare_files_fixture.py
Comment (Arcticae) [inline tests/integration/cairo_compiler/prepare_files_fixture.py:60]: Please use `Pathlib` for this

File: tests/integration/cairo_compiler/prepare_files_fixture.py
Comment (Arcticae) [inline tests/integration/cairo_compiler/prepare_files_fixture.py:41]: I don't understand why we need whole protostar project for those tests, it seems redundant

File: tests/integration/cairo_compiler/simple_contracts_compiler_test.py
Comment (Arcticae) [inline tests/integration/cairo_compiler/simple_contracts_compiler_test.py:48]: if the `PreparedFiles` entry value would be a `NamedTuple` it'd be more readable, it's hard to tell what's at 0 index right now.

**Review by Arcticae — CHANGES_REQUESTED** [review]

**Review by Arcticae — APPROVED** [review]
Actually scratch the above, i approve, but please consider fixing above nitpicks before merging, if you still have any strength left to do it 😅

File: tests/integration/cairo1/parser_test.py
Comment (karol-bisztyga) [reply tests/integration/cairo1/parser_test.py]: I don't think the location of this file is that important, the source code can be easily found and it's just additional info here.

File: tests/integration/cairo_compiler/prepare_files_fixture.py
Comment (karol-bisztyga) [reply tests/integration/cairo_compiler/prepare_files_fixture.py:60]: What do you mean?

Comment (Arcticae) [reply tests/integration/cairo_compiler/prepare_files_fixture.py:61]:
```
from pathlib import Path
Path(output_path).read_text()
```

File: tests/integration/cairo_compiler/prepare_files_fixture.py
Comment (karol-bisztyga) [reply tests/integration/cairo_compiler/prepare_files_fixture.py]: [use string based enum](https://github.com/software-mansion/protostar/pull/1316/commits/7b57752e0b001ee3892304d29b621bb5e070154e)

File: tests/integration/cairo_compiler/prepare_files_fixture.py
Comment (karol-bisztyga) [reply tests/integration/cairo_compiler/prepare_files_fixture.py]: Fixed, thanks!

File: tests/integration/cairo_compiler/prepare_files_fixture.py
Comment (karol-bisztyga) [reply tests/integration/cairo_compiler/prepare_files_fixture.py]: Fixed, thanks!

Comment (karol-bisztyga) [reply tests/integration/cairo_compiler/prepare_files_fixture.py]: Fixed, thanks!

File: tests/integration/cairo_compiler/prepare_files_fixture.py
Comment (karol-bisztyga) [reply tests/integration/cairo_compiler/prepare_files_fixture.py]: [rename RequestedFiles => RequestedFile](https://github.com/software-mansion/protostar/pull/1316/commits/f6ef81a6db163b92c7ed6c3130eaa23bec39d851)

Comment (karol-bisztyga) [reply tests/integration/cairo_compiler/prepare_files_fixture.py]: [rename RequestedFiles => RequestedFile](https://github.com/software-mansion/protostar/pull/1316/commits/f6ef81a6db163b92c7ed6c3130eaa23bec39d851)

File: tests/integration/cairo_compiler/simple_contracts_compiler_test.py
Comment (karol-bisztyga) [reply tests/integration/cairo_compiler/simple_contracts_compiler_test.py]: [use named tuple](https://github.com/software-mansion/protostar/pull/1316/commits/2c2708b1cfa062bf34a50dc6ec34f7dbba4e9e8e)

File: tests/integration/cairo_compiler/prepare_files_fixture.py
Comment (karol-bisztyga) [reply tests/integration/cairo_compiler/prepare_files_fixture.py]: `output_path` is already a `Path`

Comment (karol-bisztyga) [reply tests/integration/cairo_compiler/prepare_files_fixture.py]: [use .read_text()](https://github.com/software-mansion/protostar/pull/1316/commits/73b05b580c211cbebf1bcea5a093685ef31f6738)

File: tests/integration/cairo_compiler/prepare_files_fixture.py
Comment (karol-bisztyga) [reply tests/integration/cairo_compiler/prepare_files_fixture.py:40]: https://github.com/software-mansion/protostar/issues/1465

**Comment by karol-bisztyga** [comment]
@Arcticae @MaksymilianDemitraszek can we merge this? You can re-check if you want

**Comment by Arcticae** [comment]
@karol-bisztyga LGTM

**Comment by MaksymilianDemitraszek** [comment]
If it works I am ok with merging this

**Pull request merged**
